### PR TITLE
fix(dependabot): restore fail-closed private-index update boundary

### PR DESCRIPTION
**Pull request by Katsiarynakavaleuskaya**
## Goal

Restore the fail-closed Dependabot private-index boundary by removing external manifest-code execution authority and bounding the existing pinned requirement-carrier regex without changing its accepted grammar.

## Business reason

The private Python updater receives registry credentials. Allowing manifest code to execute expands the credential trust boundary, while the current translated requirement-line regex has a reproducible catastrophic-backtracking near-match. This PR removes that capability and closes the bounded denial-of-service path without widening dependency intake or creating a new parser.

## Scope

- Remove `insecure-external-code-execution: allow` from the sole root `pip` updater.
- Reject the key by presence in every mapping traversed by the existing bounded YAML checker, independently of its value.
- Preserve one exact private `python-index`, secret references, and `replaces-base: true`.
- Make only the overlapping whitespace repetitions and the specific overlapping numeric version repetition possessive under standard Python 3.11+ `re`.
- Add deterministic owning tests and align dependency-management prose.

## Out of scope

- No requirements, constraints, lock, workflow, private proxy, credential, Docker, RAG, frontend, iOS, OpenAPI, `AGENTS.md`, or backlog change.
- No public PyPI fallback, new carrier family, parser, tokenizer, regex dependency, policy framework, schema, API, service, crawler, suppression, skip, or contract-version bump.
- No claim that private routing proves candidate provenance, integrity, or merge readiness.

## Files changed

Exactly five material paths:

- `.github/dependabot.yml`
- `scripts/ci/check_dependabot_python_policy.py`
- `scripts/ci/dependabot_requirement_carriers.py`
- `tests/test_check_dependabot_python_policy.py`
- `docs/DEPENDENCY_MANAGEMENT.md`

## Key decisions

- Authority is reduced: absence of a safe automatic update path is `FAIL_CLOSED` and routes to the reviewed manual-update lane; restoring `allow` is not rollback.
- `FORBIDDEN_UPDATE_KEYS` remains unchanged. The existing full mapping walk is the sole owner of the static sanitized diagnostic.
- The traversal runs before semantic updater-shape early returns so every reachable mapping occurrence is reported without rendering its value or source YAML.
- `dependabot-python-requirement-carriers/v1`, its pinned upstream identity, limits, suffixes, and grammar source remain unchanged.
- Post-open validation found one distinct overlapping numeric version repetition; the operator authorized only that possessive rewrite plus one child-process regression, with no parser/framework/carrier-branch expansion.
- Any golden acceptance delta, newly admitted invalid case, second materially novel carrier, or need for a sixth material path is a stop/rescope condition.

## Tests / validation

- Red-first policy/config failures reproduced against the old `allow` contract.
- Red-first regex regressions reproduced as `subprocess.TimeoutExpired` after five seconds for the 1,000-space family and, after the authorized post-open rescope, the 10,000-digit version family.
- `check_preflight.py`: PASS.
- `check_agent_consistency.py`: PASS.
- Dependabot policy CLI: PASS.
- Owning module: PASS, 121 tests.
- Focused dependency, supply-chain, repository-policy, nosec, and absolute-subprocess guards: PASS.
- `tests/test_review_pattern_oracles.py`: PASS.
- `tests/test_agent_learning_loop.py`: PASS.
- `py_compile`: PASS.
- `make validate-changed`: PASS on the committed branch diff.
- `pre-commit run --all-files`: PASS after Black applied its normal formatting and the full bundle was rerun.
- Commit and pre-push hooks: PASS, including mypy, pip-audit, backend tests, full-repo Bandit, and Docker build test.
- `git diff --check`: PASS.
- GitHub material-head CI run 31367525279: all technical jobs PASS; the pre-closeout Merge readiness gate failed as expected before mapping publication.

## Security notes

- The forbidden diagnostic is static and contains no configured value, secret reference, arbitrary nested scalar, or source YAML.
- Representative values `allow`, `deny`, `false`, `null`, and an empty mapping are rejected by presence.
- Canonical, known nested, unknown nested dict/list, and malformed-updater early-return shapes are covered.
- One isolated `sys.executable` child rejects `!`, `@`, and `=` after 1,000 spaces; one second child rejects the authorized 10,000-digit version near-match. Both use `shell=False`, `check=True`, repository cwd, and a five-second timeout.
- Private-index routing and a successful Dependabot attempt remain intake evidence only; candidate source, version, provenance, hashes/mirror lineage, yanked/prerelease status, lock replay, drift, and targeted security/runtime gates remain review-controlled.

## Risks / rollback

- Dependabot may be unable to evaluate a dynamic manifest after authority removal. That is the intended fail-closed outcome; use the reviewed manual dependency-update lane.
- Possessive quantifiers require Python 3.11+, matching the repository 3.11/3.12/3.13 lanes. Current-head CI remains required to prove matrix compatibility.
- Rollback is a normal revert only if policy review identifies a defect; restoring external-code execution is not an acceptable operational rollback.
- No database, runtime API, dependency, lock, or migration state is involved.

## Deferred / follow-ups

- After merge, observe the next Dependabot attempt and classify only as `SUCCESS`, `FAIL_CLOSED`, or `CONFIG_ERROR`.
- A `FAIL_CLOSED` result uses the reviewed manual update lane. Candidate provenance/integrity is never inferred from this PR or an automated attempt.

## Discussion Thread Pass

- [x] Discussion-thread pass completed
- [x] Fixed in commit mapping completed

### Fixed in Commit Mapping

- [Canonical PR 2251 review mapping](https://github.com/Katsiarynakavaleuskaya/PulsePlate/blob/fix/dependabot-private-index-trust-boundary/docs/review/PR_2251_FIXED_MAPPING.md)

## Experiment Runner Evidence

- Artifact: artifacts/orchestration/experiments/results/exp-1a1f3f0d9c83-3767773b-result.json
- Apple Container oracle-only result was accepted with `mutated_paths=[]`, network isolation, and three immutable oracle commands returning zero.
- This local artifact has no review, security-provider, approval, merge, or readiness authority and did not materially shape the implementation decision.

## Lane Start Provenance

- Packet: artifacts/orchestration/task_packets/5dfca669304e.json
- Starter: scripts/orchestration/start_pr_lane.sh

## Next best step

Publish the sole mapping-only closeout commit, wait for terminal current-head checks and one review cycle, then run the strict authenticated merge-readiness wrapper before merge.

Supersedes #2242 and #2243.

## Summary by Sourcery

Запретить небезопасное выполнение внешнего кода в приватном обновляторе pip для Dependabot и усилить проверку строк с требованиями, чтобы избежать катастрофического бэктрекинга регулярных выражений, сохранив при этом неизменной допустимую грамматику.

New Features:
- Добавить проверку политики репозитория, которая сообщает о любом наличии ключа `insecure-external-code-execution` в конфиге Dependabot, независимо от его значения.
- Ввести тестовый стенд с ограничением по тайм-ауту, который прогоняет валидатор строк требований на длинных некорректных входных данных, чтобы гарантировать детерминированный отказ.

Bug Fixes:
- Предотвратить катастрофический бэктрекинг в регулярном выражении строки требований Dependabot, сделав квантификаторы пробелов обладающими свойством «possessive» в семантике `re` Python 3.11+.

Enhancements:
- Ужесточить проверку политики Dependabot, чтобы в случае необходимости выполнения внешнего кода для динамических манифестов проверка завершалась в режиме «fail closed», перенаправляя такие обновления в поток ручного рассмотрения.
- Уточнить документацию по управлению зависимостями, описав удаление полномочий на выполнение внешнего кода и подтвердив, что успешные запуски Dependabot являются лишь свидетельством принятия артефактов, а не доказательством их целостности или готовности.

Documentation:
- Обновить документацию по управлению зависимостями, чтобы отразить запрет на выполнение внешнего кода для приватного обновлятора pip и подчеркнуть необходимость ручной проверки для принятых обновлений.

Tests:
- Добавить тесты, гарантирующие, что приватный обновлятор pip больше не объявляет `insecure-external-code-execution`, что этот ключ запрещён во всех позициях отображения конфига, и что его диагностические сообщения не раскрывают настроенные значения.
- Добавить регрессионные тесты, подтверждающие, что обновлённое регулярное выражение для строк требований отклоняет длинные почти совпадающие строки с избытком пробелов за ограниченное время выполнения.

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Disallow insecure external code execution in the private pip Dependabot updater and strengthen requirement-line validation to avoid catastrophic regex backtracking while keeping the accepted grammar unchanged.

New Features:
- Add a repository policy check that reports any presence of the insecure-external-code-execution key anywhere in the Dependabot config, independently of its value.
- Introduce a timeout-guarded test harness that exercises the requirement-line validator under long invalid inputs to ensure deterministic rejection.

Bug Fixes:
- Prevent catastrophic backtracking in the Dependabot requirement-line regex by making whitespace quantifiers possessive under Python 3.11+ re semantics.

Enhancements:
- Tighten Dependabot policy validation to fail closed when external code execution would be required for dynamic manifests, directing such updates to the manual-review lane.
- Clarify dependency-management documentation to describe the removal of external code execution authority and reaffirm that successful Dependabot runs are intake evidence only, not proof of integrity or readiness.

Documentation:
- Update dependency management documentation to reflect that external code execution is now forbidden for the private pip updater and to reinforce manual review requirements for accepted updates.

Tests:
- Add tests ensuring the private pip updater no longer declares insecure-external-code-execution, that the key is forbidden at all config mapping positions, and that its diagnostic does not leak configured values.
- Add regression tests that verify the updated requirement-line regex rejects long whitespace-heavy near-matches within a bounded execution time.

</details>

## Summary by CodeRabbit

* **Security**
  * Strengthened dependency update policies by prohibiting external code execution for private Python registry updates.
  * Added fail-closed validation to prevent credential exposure through unsupported configuration settings.

* **Documentation**
  * Updated dependency management guidance with manual verification requirements for candidate updates and validation results.

* **Bug Fixes**
  * Improved requirement parsing to handle whitespace reliably and prevent ambiguous matches.

* **Tests**
  * Expanded coverage for detecting forbidden settings at all configuration levels, including nested entries.

---
## Summary by cubic
Restore the fail-closed Dependabot policy for the private Python index by forbidding external manifest code execution and bounding the requirement-line regex to prevent catastrophic backtracking. Dependabot now fails closed if a manifest requires execution.

- **Bug Fixes**
  - Reject `insecure-external-code-execution` at any mapping in `.github/dependabot.yml`; keep a single private `python-index` and `replaces-base: true`. Errors surface before shape checks and never echo values.
  - Make whitespace and numeric version runs possessive in the requirement-line regex (Python 3.11+) to block long-space and long-version near-match backtracking. Add timeout-guarded tests.

- **Documentation**
  - Update dependency policy for the fail-closed stance and bounded regex behavior.
  - Add `docs/review/PR_2251_FIXED_MAPPING.md` with review-seal dispositions.

<sup>Written for commit 964bdcbad110b5f2da8fa979ddf4a1f2e6dd4a84. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/Katsiarynakavaleuskaya/PulsePlate/pull/2251?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: scripts/ci/dependabot_requirement_carriers.py [hunk 1]
```diff
@@ -86,14 +86,19 @@ class DependabotRequirementCarrierUpstreamSnapshot:
...
     rf"\s*\\?\s*\(?(?:{_UPSTREAM_REQUIREMENTS})?\)?"
     rf"\s*\\?\s*(?:;\s*{_UPSTREAM_MARKER_EXPRESSION})?"
     rf"\s*\\?\s*(?:{_UPSTREAM_HASHES})?"
-    r"\s*(?:#+\s*.*)?$",
+    r"\s*(?:#+\s*.*)?$"
+)
+# Whitespace in this grammar separates tokens. Possessive repetitions prevent
+# adjacent optional branches from redistributing the same whitespace run.
+_UPSTREAM_VALID_REQUIREMENT_LINE_RE = re.compile(
+    _UPSTREAM_VALID_REQUIREMENT_LINE_PATTERN.replace(r"\s*", r"\s*+").replace(r"\s+", r"\s++"),
```
[the remark this answers is not in the document]
Comment (Katsiarynakavaleuskaya) [reply scripts/ci/dependabot_requirement_carriers.py]: Disposition: FIXED
Commit: b0efc2d3cb2f40477b084c76b81b5ba55841058c
Evidence: scripts/ci/dependabot_requirement_carriers.py:98-108; tests/test_check_dependabot_python_policy.py:229; owning suite 121 PASS; exact-material-head CI run 31367525279 test-pr PASS.
Mapping: docs/review/PR_2251_FIXED_MAPPING.md.

File: scripts/ci/dependabot_requirement_carriers.py [hunk 2]
```diff
@@ -86,14 +86,19 @@ class DependabotRequirementCarrierUpstreamSnapshot:
...
     rf"^\s*\\?\s*{_UPSTREAM_NAME}"
     rf"\s*\\?\s*(?:\[\s*{_UPSTREAM_EXTRA}"
     rf"(?:\s*,\s*{_UPSTREAM_EXTRA})*\s*\])?"
     rf"\s*\\?\s*\(?(?:{_UPSTREAM_REQUIREMENTS})?\)?"
     rf"\s*\\?\s*(?:;\s*{_UPSTREAM_MARKER_EXPRESSION})?"
     rf"\s*\\?\s*(?:{_UPSTREAM_HASHES})?"
-    r"\s*(?:#+\s*.*)?$",
+    r"\s*(?:#+\s*.*)?$"
+)
+# Whitespace in this grammar separates tokens. Possessive repetitions prevent
```
[the remark this answers is not in the document]
Comment (Katsiarynakavaleuskaya) [reply scripts/ci/dependabot_requirement_carriers.py]: Disposition: NOT-A-BUG
Evidence: .github/workflows/ci.yml:8,611-613; scripts/ci/dependabot_requirement_carriers.py:101-109; Python 3.13.14 compilation plus owning suite 121 PASS and exact-material-head CI run 31367525279 lint/test-pr PASS.
Reason: standard re supports possessive quantifiers on the repository Python 3.11+ floor. A helper or generalized parser owner would violate the frozen narrow contract.
Mapping: docs/review/PR_2251_FIXED_MAPPING.md.

**Comment by Katsiarynakavaleuskaya** [comment]
Superseded by #2260
